### PR TITLE
Add Introduction to Logging changes in ASP.NET Core 2.0

### PR DESCRIPTION
**Pull request by pakrym**
/cc @Rick-Anderson  @DamianEdwards @glennc

### CODE REVIEW

File: aspnetcore/fundamentals/logging-2.0.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                })
+                .ConfigureLogging((hostingContext, logging) =>
+                {
+                    logging.AddConfiguration(hostingContext.Configuration.GetSection("Logging"));
+                    logging.AddConsole();
+                    logging.AddDebug();
+                });
+```
+
+A big advantage shown in this example is using the `HostingContext` provided to the `ConfigureLogging` extension method. This allows easy access to the configuration that is also registered in `Program.cs`. This leaves your `Startup.cs` more focused on configuring services and your middleware pipeline with less other concerns (namely logging and configuration).
```
Comment (ardalis) [inline aspnetcore/fundamentals/logging-2.0.md]: "with less other concerns" -> "with fewer other concerns"

File: aspnetcore/fundamentals/logging-2.0.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+The final thing that impacts configuring logging is the new opinionated `WebHost` that you can see in the ASP.NET Core 2.0 templates. The default `WebHost` will configure logging in the same way as the sample above, meaning that if you only need `Console` and `Debug` logging and will use a `Logging` configuration section to control things like MinimumLogLevel or filters, then you don't need to add any extra code. If you want to add additional `ILoggerProviders` then you can call `ConfigureLogging` as well, adding just the providers that you need:
+
+```csharp
+WebHost.CreateDefaultBuilder(args)
+    .UseStartup<Startup>()
+    .ConfigureLoggin(logging => logging.AddEventLog())
```
Comment (ardalis) [inline aspnetcore/fundamentals/logging-2.0.md]: ConfigureLoggin -> ConfigureLogging

File: aspnetcore/fundamentals/logging-2.0.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    .Build();
+```
+In the above sample you have all of the default loggers (Console and Debug) as well as the EventLog provider that you are adding.
+
+## Configuring Filtering
+
+In ASP.NET Core 1.x there are ways to do filtering, but they are either not very intuitive or are limited to specific log providers. In 2.0 we have changed the default `ILoggerFactory` to support filtering, meaning that you can filter log messages to all providers, and allow binding the filtering to `IConfiguration` allowing you to control it from whatever configuration sources you choose, such as a file.
```
Comment (ardalis) [inline aspnetcore/fundamentals/logging-2.0.md]: Sentence is too long:
"to support filtering. This means that..." 
"and allows binding the filtering to `IConfigurationg`."
"This new approach lets you control filtering from whatever configuration sources you choose (for example, a settings file)."

File: aspnetcore/fundamentals/logging-2.0.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+}
+```
+
+The data you have available when filtering log messages is:
+- Provider type/alias
+- Category name
+- Minimum level
+- Filter function
+
+You can apply a filter function to a specific provider or category as well as globally for all providers and categories. When registering a global filter function you can get access to the provider, category, and LogLevel to decide whether or not a messages should be logged. For example:
```
Comment (ardalis) [inline aspnetcore/fundamentals/logging-2.0.md]: "a messages" -> "a message"

File: aspnetcore/fundamentals/logging-2.0.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+Getting providers from DI and using LoggerFilterOptions to filter messages are both features of the default LoggerFactory provided by Microsoft. If you want to replace the factory you can do so by replacing the `ILoggerFactory` service in DI:
+
+```csharp
+ConfigureServices(collection => collection.AddSingleton<ILoggerFactory>(myFactory))
+```
+
+## Filtering Algorithm
+
+When creating an `ILogger` to write logs to, we select a single rule per provider to apply to this logger. All messages written to the `ILogger` are filtered based on these rules. We select the most specific rule possible for each provider and category pair from the available filter.
```
Comment (ardalis) [inline aspnetcore/fundamentals/logging-2.0.md]: "to write logs to" -> "to log messages to"

File: aspnetcore/fundamentals/logging-2.0.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+The following algorithm is used to select a rule for a given provider and category:
+
+1. Select rules that match provider or its alias, if none found - select all rules with an empty provider.
+2. From result of 1 select rules with longest matching category prefix, if none found - select all rules without category specified.
+3. If multiple rules got selected take **last** one
+4. If no rules selected use MinimumLevel for a level
+
+NOTE: Filter function is only invoked when message level matches rule minimum level
```
Comment (ardalis) [inline aspnetcore/fundamentals/logging-2.0.md]: This algorithm really needs several examples to demonstrate how it works.

**Comment by Rick-Anderson** [comment]
@pakrym  can you hold off on new commit until I complete my PR #3604 on this commit?

**Comment by pakrym** [comment]
@Rick-Anderson sure

**Comment by tdykstra** [comment]
I'm taking this content into a new PR to the 2.0 Release Notes doc, and later to the existing logging doc.

**Pull request closed without merging**